### PR TITLE
Clean halo depth

### LINKED ISSUE #329: Get clean halo depth
### Are there any linked Issues or Pull Requests?

PR for this issue: #333 

### Brief description

Add functionality to a field to return the depth of the field's halos that are marked as clean.

### Further details of the issue.

PR #303 optimises the calculation of PSML by performing redundant calculations into the halos to reduce the number of halo exchanges required. It does this through a kernel that uses the ``operates_on`` metadata of ``owned_and_halo_cell_column``. This is an inefficient way to perform this task as the kernel is column-based and it would be quicker just to do a dof-based loop through the field data, out to the required halo depth.

Andy Porter has come up with the suggestion of making the built-in ``setval_X`` loop out to the last clean halo depth (by default - optimisation scripts could limit this to other values like last owned dof or last annexed dof). He has documented this in stfc/PSyclone#3398 which would support the PSML optimisation in #348.

For Andy's suggestion to work, the infrastructure needs to be able to return the depth of halos that are marked as clean. It currently can't do this, so it needs adding to the infrastructure.

### PR DESCRIPTION
**Pull request by mike-hobson**
# PR Summary

Sci/Tech Reviewer: @MatthewHambley 
Code Reviewer: @EdHone 

In order to develop new functionality in PSyclone, STFC require an additional function adding to the field proxy that returns how many "depths of halo" are clean. I.e. if the first halo to be marked as dirty is at depth 3, then the new function (``get_clean_depth()``) returns 2.

A minor complication is that we sometimes call the annexed dofs the "zeroth" halo, so returning a clean depth of ``0`` could either mean there are no clean halos or that the annexed dofs are clean. To eliminate this confusion the new function returns a larger negative number as a return code to indicate that all halos are dirty.

There are new unit tests for the new functionality. When writing the tests, I noticed that the test for ``get_last_dof_halo(0)`` is wrong - it checks the returned value against ``last_owned_dof``, whereas it should be checking against ``last_annexed_dof``. This happens to work because we currently only test function spaces in serial where ``last_owned_dof == last_annexed_dof``. So, a) I have corrected the incorrect unit test and b) To catch this in the future, I have opened issue #332 to parallelise the function space unit tests (this task is too big for this PR).

closes #329 

## Code Quality Checklist

- [x] I have performed a self-review of my own code
- [x] My code follows the project's [style guidelines](https://metoffice.github.io/lfric_core/how_to_contribute/index.html#how-to-contribute-index)
- [x] Comments have been included that aid understanding and enhance the readability of the code
- [x] My changes generate no new warnings
- [ ] All automated checks in the CI pipeline have completed successfully

## Testing

- [x] I have tested this change locally, using the LFRic Core rose-stem suite
- [ ] If required (e.g. API changes) I have also run the LFRic Apps test suite using this branch
- [ ] If any tests fail (rose-stem or CI) the reason is understood and acceptable (e.g. kgo changes)
- [x] I have added tests to cover new functionality as appropriate (e.g. system tests, unit tests, etc.)
- [x] Any new tests have been assigned an appropriate amount of compute resource and have been allocated to an appropriate testing group (i.e. the developer tests are for jobs which use a small amount of compute resource and complete in a matter of minutes)

# Test Suite Results - lfric_core - clean_halo_depth/run2

## Suite Information

| Item | Value |
| :--- | :--- |
| Suite Name | [clean_halo_depth/run2](https://cylchub/services/cylc-review/cycles/mike.hobson/?suite=clean_halo_depth%2Frun2) |
| Suite User | mike.hobson |
| Workflow Start | 2026-04-16T10:07:13 |
| Groups Run | all |

| Dependency | Reference | Main Like |
| :--- | :--- | :--- |
| lfric_core | [mike-hobson/lfric_core@clean_halo_depth](https://github.com/mike-hobson/lfric_core/tree/clean_halo_depth) | False |
| SimSys_Scripts | [MetOffice/SimSys_Scripts@4387949](https://github.com/MetOffice/SimSys_Scripts/tree/4387949) | True |

## Task Information
:white_check_mark: succeeded tasks - 390

## Security Considerations

- [ ] I have reviewed my changes for potential security issues
- [ ] Sensitive data is properly handled (if applicable)
- [ ] Authentication and authorisation are properly implemented (if applicable)

## Performance Impact

- [x] Performance of the code has been considered and, if applicable, suitable performance measurements have been conducted

## AI Assistance and Attribution

- [ ] Some of the content of this change has been produced with the assistance of _Generative AI tool name_ (e.g., Met Office Github Copilot Enterprise, Github Copilot Personal, ChatGPT GPT-4, etc) and I have followed the [Simulation Systems AI policy](https://metoffice.github.io/simulation-systems/FurtherDetails/ai.html) (including attribution labels)

## Documentation

- [x] Where appropriate I have updated documentation related to this change and confirmed that it builds correctly

## PSyclone Approval

- [ ] If you have edited any PSyclone-related code (e.g. PSyKAl-lite, Kernel interface, optimisation scripts, LFRic data structure code) then please contact the [TCD Team](mailto:ToolsCollabDevTeam@metoffice.gov.uk)

# Sci/Tech Review

- [x] I understand this area of code and the changes being added
- [x] The proposed changes correspond to the pull request description
- [x] Documentation is sufficient (do documentation papers need updating)
- [x] Sufficient testing has been completed

(_Please alert the code reviewer via a tag when you have approved the SR_)

# Code Review

- [ ] All dependencies have been resolved
- [ ] Related Issues have been properly linked and addressed
- [ ] CLA compliance has been confirmed
- [ ] Code quality standards have been met
- [ ] Tests are adequate and have passed
- [ ] Documentation is complete and accurate
- [ ] Security considerations have been addressed
- [ ] Performance impact is acceptable

### CODE REVIEW

File: infrastructure/source/field/field_parent_mod.f90 [hunk 1]
```diff
@@ -489,6 +493,28 @@ subroutine set_clean(self, depth)
...
+
+    class(field_parent_proxy_type), intent(in) :: self
+
+    integer(i_def) :: depth
+
+    if (self%halo_dirty(1) == 1) then
+      depth = all_halos_dirty
+    else
+      do depth=1,self%field_halo_depth-1
+        if (self%halo_dirty(depth+1) == 1) exit
```
Comment (MatthewHambley) [inline infrastructure/source/field/field_parent_mod.f90]: This is a clever trick to take advantage of already having tested `self%halo_dirty(1)` and to ensure `depth` holds the correct value. It could be easily overlooked or misunderstood so a comment to explain what's going on might be a good idea.

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
This is a short and straight forward change which looks good to me. I have only one comment to consider before giving my approval.

**Review by MatthewHambley — APPROVED** [review]
This looks to be in good shape and is ready for code review.

File: infrastructure/source/field/field_parent_mod.f90 [hunk 2]
```diff
@@ -489,6 +493,30 @@ subroutine set_clean(self, depth)
...
+
+    integer(i_def) :: depth
+
+    ! If the first halo depth is dirty return the "all dirty" flag.
+    ! Otherwise, start at depth 2 looking for the first dirty halo depth
+    if (self%halo_dirty(1) == 1) then
+      depth = all_halos_dirty
+    else
+      do depth=1,self%field_halo_depth-1
+        if (self%halo_dirty(depth+1) == 1) exit
```
Comment (EdHone) [inline infrastructure/source/field/field_parent_mod.f90]: I'm not sure what would happen for the case of a field with a clean halo with depth 1. Looking at the code here I think it would return uninitialised data, which wouldn't be ideal. I think you might need some more checks in your loop, and maybe some more test cases in your unit test would help to ensure that the behaviour is as expected for all cases.

**Review by EdHone — CHANGES_REQUESTED** [review]
Just the one question regarding the completeness of the logic in the `get_clean_depth` routine.

File: infrastructure/source/field/field_parent_mod.f90 (hunk 2 above)
Comment (mike-hobson) [reply infrastructure/source/field/field_parent_mod.f90]: Hi Ed, Thanks for looking at that. I've had a look through the code and the only thing I can see that's wrong with the logic is that it can't cope when there are no halos, so I've put in a check for that.

Otherwise:

1. if the first halo depth is dirty, depth is always set to ``all_halos_dirty`` no matter what the halo depth is.
2. if the first halo is clean and the halo depth is 1, the ``depth`` loop counter will be set to 1, but it doesn't execute the contents of the loop - so a depth of 1 is returned.
3. if the first halo is clean and the halo depth is 2 (with the second halo, dirty), the ``depth`` loop counter is set to 1, and the contents of the loop is executed once. The ``exit`` will be hit and a depth of 1 will be returned
4. if two halos are clean and the halo depth > 2 (with the remaining halos, dirty),  the contents of the loop is executed until it finds the dirty halo and hits the ``exit``. This will return a depth that is 1 less than the first dirty halo - i.e. the last clean halo.
5. If all the halos are clean, the loop will be executed without ever hitting the ``exit`` until it reaches ``self%field_halo_depth``, so returns the full depth of the halos.

To help prove all of this works, I've added unit tests for a field with a single depth halo that is clean, a test for when all halos are clean and a test for when there are no halos. 

Is that ok?

Comment (EdHone) [reply infrastructure/source/field/field_parent_mod.f90]: Thanks, Mike. With the extra edge-cases tested I'm happy to approve. Your explanation above makes sense and helped me to understand a little more about how loops are executed in these particular cases!

**Review by EdHone — APPROVED** [review]
All changes approved and tests running successfully

**Review by EdHone — APPROVED** [review]
